plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+// See the file LICENSE for licensing terms.
+
+// ACP176 implements the fee logic specified here:
+// https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/176-dynamic-evm-gas-limit-and-price-discovery-updates/README.md
+package acp176
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+	"sort"
+
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/holiman/uint256"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+const (
+	MinTargetPerSecond  = 1_000_000                                 // P
+	TargetConversion    = MaxTargetChangeRate * MaxTargetExcessDiff // D
+	MaxTargetExcessDiff = 1 << 15                                   // Q
+	MinBaseFee          = 1                                         // M
+
+	TimeToFillCapacity            = 10   // in seconds
+	TargetToMax                   = 2    // multiplier to convert from target per second to max per second
+	TargetToPriceUpdateConversion = 43   // 43s ~= 30s * ln(2) which makes the price double at most every ~30 seconds
+	MaxTargetChangeRate           = 1024 // Controls the rate that the target can change per block.
+
+	targetToMaxCapacity = TargetToMax * TimeToFillCapacity
+	maxTargetExcess     = 1_024_950_627 // TargetConversion * ln(MaxUint64 / MinTargetPerSecond) + 1
+)
+
+var minBaseFee = big.NewInt(MinBaseFee)
+
+// State represents the current state of the gas pricing and constraints.
+type State struct {
+	Gas          gas.State
+	TargetExcess gas.Gas
+}
+
+// Target returns the target gas consumed per second.
+//
+// Target = MinTargetPerSecond * e^(TargetExcess / TargetConversion)
+func (s *State) Target() gas.Gas {
+	return gas.Gas(gas.CalculatePrice(
+		MinTargetPerSecond,
+		s.TargetExcess,
+		TargetConversion,
+	))
+}
+
+// MaxCapacity returns the maximum possible accrued gas capacity.
+func (s *State) MaxCapacity() gas.Gas {
+	targetPerSecond := s.Target()
+	maxCapacity, err := safemath.Mul(targetToMaxCapacity, targetPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	return maxCapacity
+}
+
+// BaseFee returns the current required fee per gas.
+//
+// BaseFee = MinBaseFee * e^(Excess / (Target() * TargetToPriceUpdateConversion))
+func (s *State) BaseFee() *big.Int {
+	target := s.Target()
+	priceUpdateConversion, err := safemath.Mul(TargetToPriceUpdateConversion, target)
+	if err != nil {
+		priceUpdateConversion = math.MaxUint64
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
```

Comment:
Why is this different to the `MaxCapacity()` method? They both look like $C$ to me.

Reply:
This is also generating `MaxCapacity` - it just felt wasteful to call `Target()` again (as that is much more expensive than just `safemath.Mul`.